### PR TITLE
Add pretty printing

### PR DESCRIPTION
**Pull request by melund**
The pretty print module is from the IPython project. Does anyone know how to give proper credit etc...

### CODE REVIEW

**Comment by asmeurer** [comment]
If you're going to use a custom thing instead of `pprint` you might as well make the xonsh custom classes pretty print as well.

File: xonsh/pretty.py [hunk 1]
```diff
@@ -0,0 +1,847 @@
+# -*- coding: utf-8 -*-
+"""
+Python advanced pretty printer.  This pretty printer is intended to
+replace the old `pprint` python module which does not allow developers
+to provide their own pretty print callbacks.
+
+This module is based on ruby's `prettyprint.rb` library by `Tanaka Akira`.
+
```
Comment (scopatz) [inline xonsh/pretty.py]: Add the IPython copyright notice here on this line, please.  See https://github.com/scopatz/xonsh/blob/master/xonsh/tools.py#L3 for an example.

**Comment by scopatz** [comment]
Thanks for putting this in @melund!

> If you're going to use a custom thing instead of pprint you might as well make the xonsh custom classes pretty print as well. 

I agree.  In particular Env and Aliases are most often touched.

Also please add a hook in the docs for this new module.  Should be easy, just follow how the other modules autodoc.

**Comment by melund** [comment]
@scopatz . I added a copyright notice, documentation.

Pretty print of Aliases look like this now (but can be customized)...

``` 
mel@ANYBODY-TRAVEL1 ~\Documents\GitHub\xonsh pretty_print $ aliases
xonsh.built_ins.Aliases(
{'EOF': <function xonsh.aliases.exit>,
 'activate': <function _activate>,
 'bg': <function xonsh.jobs.bg>,
 'cd': <function xonsh.dirstack.cd>,
 'cls': ['cmd', '/c', 'cls'],
 'copy': ['cmd', '/c', 'copy'],
 'd': ['cmd', '/c', 'dir'],
 'deactivate': <function _deactivate>,
 'del': ['cmd', '/c', 'del'],
 'dir': ['cmd', '/c', 'dir'],
....
```

**Comment by scopatz** [comment]
Thanks @melund! This is very much appreciated!

**Pull request merged**